### ISSUE TITLE
Parameter optimization [feature request]

### ISSUE THREAD

**Issue by sjackman** [nextflow-io/nextflow#168]
I would like a tool to optimize the parameters of an arbitrary command line program for a variable output by that tool (on its standard output, say). For example, I want to run genome assemblies varying the value of the parameter _k_ to optimize the _N50_ output by the assembler. It's easy to do this optimization now using a grid search, running the tool for all possible values of that parameter, but I'd like to something a bit smarter and run fewer assemblies. i.e. assemble at _k_=64, 128, 192 and based on the results of those assemblies pick new values of _k_ at which to assemble. The real power would come in optimizing multiple variables. Is optimization a feature that you would consider implementing in Nextflow?

**Comment by sjackman** [comment]
I haven't yet tried this tool, but for an example that is not integrated with the job scheduler, see https://github.com/sseemayer/ParOpt#readme

**Comment by pditommaso** [comment]
This sounds quite a cool proposal. However I need to understands if it can fit in the nextflow model. How would you define the performance metrics in this context. It would be enough to a capture one or more outputs produced by the tool? 

Also once you learned the optimal parameters how would you apply them? You would just use the output produced by the tool with the optimal parameters or you want to re-execute the tool with a super-set of that params one ore more time?

**Comment by sjackman** [comment]
> How would you define the performance metrics in this context.

The target rule being optimized would output the metric on standard output.

> Also once you learned the optimal parameters how would you apply them?

I'd be inclined to save the results of all the jobs run so far in a directory structure based on the parameter names and values.

**Comment by sjackman** [comment]
I was considering proposing a topic like this for the Hackseq hackathon in Vancover before ASHG 2016 in October. See http://www.hackseq.com . I hadn't decided on a platform yet. I wanted to check in with you to see if you thought it was a good fit for Nextflow. Any chance you'll be in Vancouver, Canada for ASHG 2016? Abstract deadlines are tomorrow for both ASHG 2016 and Hackseq.

**Comment by pditommaso** [comment]
Yes, I think it could definitively fit in nextflow. Maybe having something in DSL like this: 

    process foo { 
      input: 
      file bam from .. 
      
      optimise: 
      val x from range( .. ) 
      val y from range( .. ) 

      """ 
      tool-cmd-line $x $y $bam
      """ 
    }


Though it's unlikely I can work on it in the short run, I will try to give a try to it at my first convenience and surely I can advice if you manage to propose it for the Hackseq. 

Unfortunately I think I have very low chances to join this hackathon. I'm going to the OBF Codefest and BOSC conference in Orlando this year and this make very difficult I can also come to Vancouver in October. But I will take seriously in consideration for next year.

**Comment by pditommaso** [comment]
Alternatively if you could manage to isolate the optimisation logic in a running Java (or Groovy) skeleton, I could try to integrate it in nextflow much more quickly. 

It doesn't need to be a parallel implementation nor to execute real tasks. It should only implement the optimisation logic. What do you think?

**Comment by sjackman** [comment]
I'm not particularly Java nor Groovy savvy. I'd probably implement it in one of C++, Python, Ruby or R. I was going to start by trying out [ParOpt](https://github.com/sseemayer/ParOpt) and see if it more or less does what I need. I was going to start by having it submit just one job at a time to the scheduler, and block until that job is complete. ParOpt is only 236 lines of Python, so the optimization logic itself is not that complex. Managing the scheduler and parallel executing will be the trickier task, as I'm sure you could guess.

**Comment by pditommaso** [comment]
OK, I will to give a look to the Python code if so.

**Comment by pditommaso** [comment]
I was giving a try to this feature request and I'm starting to think that in some extent it's already possible to run such optimisation with nextflow. At the end of the day what you are proposing is to run a command multiple times over a finite set of parameters and then pick the best result. 

Currently you could do something like this:

```
process foo {
  input: 
  file 'reads.fa' from reads 
  each k from 64, 128, 192
  
  output: 
  file 'k*/contigs.fa' into contigs 
  
  """
  abyss-pe -C k$k in=reads.fa
  """
}
```

This run multiple assemblies for each `k` in the specified range (and using the same read file). Then of course it still required to pick the best result that can be done with a separated process. Note that when specifying multiple `each` statements a new task is executed for each combination of the provided values. Thus, if I'm not wrong, it's the same behaviour of the grid search. 

In the light of this, can we say that  the goal of this feature request is to have a built-in mechanism in the `process` scope that allows one to pick the best results given a set of parameters, instead of having to use a separate process/logic? Does make sense?

**Comment by sjackman** [comment]
Nice example, by the way. :wink: Yes, Nextflow already implements a grid search, and a second rule `bar` could depend on the output of `foo` and select the optimal value of *k*.

Where it gets fun is with multiple parameters, say *s* and *n*. A grid search would examine all possible combinations of *s* and *n*. A line search would pick a value for *s*, optimize *n*, then optimize *s*, and go back and forth until the stopping condition. Even that simple algorithm would be a good start, and it's easy to extend to multiple parameters. Two other multivariate optimization algorithms that are easy to describe and implement are [twiddle](https://www.youtube.com/watch?v=2uQ2BSzDvXs) and [amoeba](https://en.wikipedia.org/wiki/Nelder–Mead_method).

Even in the univariate case there's room for improvement. Let's say for example the peak NG50 occurs at *k*=152. Rather than test all values of *k* between 1 and 256, it could use a binary-search-like pattern that runs the following search pattern, with the best NG50 shown at each step in bold:

1. 0 **128** 256
2. 64 **128** 192
3. 96 128 **160**
4. **144** 160 176
5. 136 144 **152**

It finds the optimal value for *k* within a tolerance of 8 using 10 assemblies, whereas a grid search requires 32 assemblies.

**Comment by sjackman** [comment]
Any thoughts on including some simple optimization in Nextflow? It's okay if the answer is not a priority.

**Comment by pditommaso** [comment]
Well, after 4 years this issue was opened replying it's not a priority I don't think it's a reasonable answer 😬. 

The reality is this is one of my favourite enhancement for the NF language but I haven't any time to look at this so far. I really encourage the community to draft a proposal. Happy to give advice on a possible implementation.

**Comment by abhi18av** [comment]
I have explored using Nextflow for a [grid-search process in the context of ML](https://github.com/abhi18av/nextflow_grid_search/) however, it relies completely on the facilities provided by the wrapped tool, in this case the [H2O.ai library](https://docs.h2o.ai/h2o/latest-stable/h2o-docs/grid-search.html#supported-grid-search-hyperparameters).

I am guessing that to implement this within Nextflow, there needs to be some kind of a feedback provided by the tool, in the case of ML it's a metric like AUC/Accuracy etc. We could always explicitely add a `sed/grep` script to the process via `afterScript` and use that for the feedback, this is inspired from the way [nf-core is extracting `version` info](https://github.com/nf-core/modules/blob/e02f0de5146f85de4a42e3505ba91f3488158d6a/software/picard/markduplicates/main.nf#L43) from a process 

In addition to this metric, syntactically, we could  have something like `optimize` option for the numeric basic param types like `integer`, `floats` and `strings`.


```
process foo {
  input: 
  file 'reads.fa' from reads 
  k from 64, 128, 192, optimize: true
  j from "a", "b", "c", optimize: true

  output: 
  file 'k*/contigs.fa' into contigs 

  """
  abyss-pe -C k$k in=reads.fa
  """
 
 afterScript  'SOME_BASH_COMMAND', optimization: 'max', strategy: 'brute_force'

}

```


Then based on the variables `j` and `k` and the output of `afterScript`, Nextflow could selectively launch newer processes as per the grid search strategy. To begin with, we could a brute force strategy and only have `foo` with maximum value of `afterScript` as the final version. Later on, we could add other `strategy` options such as `random` or `evolutionary`.

I could be off the tempo here, but this is just a suggestion for this feature.

**Comment by pditommaso** [comment]
What the command in the `afterScript` does in your example? compute the score function?

**Comment by abhi18av** [comment]
Yes, it needs to output a comparable value for example, a number. This addition allows the users to explicitely specify a feedback for any process.

**Comment by pditommaso** [comment]
Then, it could make sense to add a new component named `scoreFunction` or something similar, that can be either a groovy/java code or a Bash script for external command and NF could call after each task execution.

**Comment by abhi18av** [comment]
Paolo, when you say `component` - does it signify a new `directive` or as an option to the present `afterScript` ?

**Comment by pditommaso** [comment]
Yes

**Comment by abhi18av** [comment]
Cool, I'd like to try and implement this feature. This seems like to a good opportunity to understand the DSL level implementation as well.

**Comment by sjackman** [comment]
Amazing! In this comment https://github.com/nextflow-io/nextflow/issues/168#issuecomment-223765318 above I suggested line search, grid search, [twiddle](https://youtu.be/2uQ2BSzDvXs), and amoeba ([Nelder–Mead](https://en.wikipedia.org/wiki/Nelder–Mead_method)) as straight forward optimization algorithms to start with, though there's definitely more powerful algorithm. [Kriging](https://en.wikipedia.org/wiki/Kriging) (Bayesian Optimization with Gaussian Processes) is powerful for small numbers of trials (say less than a hundred), since without heuristics the complexity increases the square of the number of trials,\.

**Comment by pditommaso** [comment]
The ability to have the optimisation algorithm would be very nice. The challenging is how to manage the parallelism since I have the impression that most of these algorithms are based on a feedback loop iteration to determine the descended function and minimize the objective function.  Is that correct? In that case, I don't how we could improve the parallelization.

**Comment by abhi18av** [comment]
Yeah, that sounds reasonable. However, we can begin by implementing the most basic scoringFunction and could simply rely on exhaustive search and launch the processes in the search space, collect the results separately , similar to how each repeater is supposed to work.

Then, we start iterating and improving this feature.

**Comment by sjackman** [comment]
> The ability to have the optimisation algorithm would be very nice. The challenging is how to manage the parallelism since I have the impression that most of these algorithms are based on a feedback loop iteration to determine the descended function and minimize the objective function. Is that correct? In that case, I don't how we could improve the parallelization.

Grid search and line search are embarrassingly parallel. Twiddle can be parallelized by running the 8 adjacent points in parallel, for the 2-D case. Nelder–Mead could be parallelized by having multiple amoeba running around, and merging them if they converge, which would be a super fun visualization, and makes the algorithm more robust to getting stuck in local minima. Kriging can be parallelized by running say the best three experiments, rather than just the best experiment, with each each iteration. I feel that many optimization algorithms can be parallelized with some adaptation of the algorithm.

**Comment by pditommaso** [comment]
OK, let's prototype the implementation for the linear search. My understanding is that it should work like this 

1. define a range of values against which the task should be optimised
2. *n* task executions are run in parallel 
3. each time a task complete, the score function is evaluated. 
4. if the result of the score function is lower than a previous execution the task index, inputs and outputs are retained, otherwise discarded. 
5. when the process complete (all tasks have been executed) the retained output are emitted by the process as the optimal result. 

Does it make sense?

**Comment by abhi18av** [comment]
Yeah, it makes sense 👍 

I thought about the implementation a bit more and at syntactic level it seems that we could only focus on adding the `scoringFunction` directive and focus on its options - which allows us to leave everything else in the DSL as it is.

Here's a snippet 

```nextflow

scoringFunction 'BASH_SCRIPT', variables: [i, j, k], optimization: 'max', strategy: ...

```

The `BASH_SCRIPT` and the `variables`  are the fields which need to be mentioned, rest can have defaults like  

```groovy
optimization: 'max', // default 
strategy: 'brute_search', // default
maxGridSearch: 10 // default is to launch all tasks and cover entire search_space
```

**Comment by bentsherman** [comment]
Stumbled upon this thread while browsing open PRs and wanted to add my thoughts, since this issue is relevant to ML applications.

A brute force search can already be implemented in Nextflow -- use `each` to generate all of the trials, save the output metric(s), and use channel logic to pick the best trial. Especially with DSL2 a grid search should be straightforward.

An iterative search (i.e. binary search) requires you to run one trial at a time, picking new inputs after each iteration. For this kind of search Nextflow needs to support feedback loops, which is currently in the works.

Between those two cases, I don't see why we need to add new syntax such as `optimize` or `score` keywords.  I suppose they could be sugar over the base pattern, but better I think to just use the pattern itself and have examples on the nextflow patterns page. Let me know if I'm missing something
